src/components/AppTodoContext/AppTodoContext.tsx

```diff
@@ -0,0 +1,58 @@
+import React, { FC, useMemo, useState } from 'react';
+import { Todo } from '../../types/Todo';
+import { Props, Value } from './AppTodoContext.types';
+import { ErrorType } from '../Error/Error.types';
+
+export const AppTodoContext = React.createContext<Value>({
+  todos: [],
+  setTodos: () => {},
+  visibleTodos: [],
+  setVisibleTodos: () => {},
+  tempTodo: null,
+  setTempTodo: () => {},
+  processingTodoIDs: [],
+  setProcessingTodoIDs: () => {},
+  completedTodos: [],
+  uncompletedTodos: [],
+  errorMessage: 'No error' as ErrorType,
+  setErrorMessage: () => {},
+});
+
+export const AppTodoProvider: FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [processingTodoIDs, setProcessingTodoIDs] = useState<number[]>([]);
+  const [visibleTodos, setVisibleTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState<ErrorType>(
+    ErrorType.NoError,
+  );
+
+  const completedTodos = useMemo(() => {
+    return todos.filter(todo => todo.completed);
+  }, [todos]);
+
+  const uncompletedTodos = useMemo(() => {
+    return todos.filter(todo => !todo.completed);
+  }, [todos]);
```

Comment:
O(2n) -> O(n)
```suggestion 
  const {
    completedTodos,
    uncompletedTodos,
  } = useMemo(() => {
    return todos.reduce((acc, todo) => ({
    	completedTodos: todo.completed
    	  ? [...acc.completedTodos, todo],
    	  : acc.completedTodos
    	uncompletedTodos: !todo.completed
    	  ? [...acc.uncompletedTodos, todo],
    	  : acc.uncompletedTodos
    }), {
      completedTodos: [],
      uncompletedTodos: [],
    });
  }, [todos]);
```

or

```suggestion 
  const {
    completedTodos,
    uncompletedTodos,
  } = useMemo(() => {
    return todos.reduce((acc, todo) => {
    	if (todo.completed) {
    	  acc.completedTodos.push(todo)
    	} else {
    		acc.uncompletedTodos.push(todo)
    	}
    	
    	return acc;
    }, {
      completedTodos: [],
      uncompletedTodos: [],
    });
  }, [todos]);
```

Reply:
I've tried, it brokes all :) should i come on QnA?